pandas/core/indexes/api.py

```diff
@@ -89,6 +89,19 @@ def get_objs_combined_axis(
     Index
     """
     obs_idxes = [obj._get_axis(axis) for obj in objs]
+    if all_indexes_same(obs_idxes):
```

Comment:
this is not the place to do at all. should be in concat itself.

ideally we want a nice error message if this happens (and the indices are not equal).

Reply:
I agree I need to add the error message, but I do think this function is where the error should be thrown.

Every function which uses `get_objs_combined_axis` throws an error if there are duplicate indices. Here is a demonstration of this from pandas 1.1.5:

<details>

```python
import numpy as np
import pandas as pd
import pandas._testing as tm

def list_of_series_constructor(*args):
    return pd.DataFrame(list(args))

def crosstab(*args):
    return pd.crosstab(*args)

def concat_series(*args):
    return pd.concat(list(args), axis=1)
    
def concat_dfs(*args):
    return pd.concat([pd.DataFrame(x) for x in args], axis=1)

def duplicated_indices_errors(func, index_makers):
    records = []
    for i, index_maker in enumerate(index_makers):
        rec = {"index_maker": index_maker.__name__}
        uniq = index_maker(k=4)
        non_uniq = uniq[[0, 0, 1, 2]]

        s_uniq = pd.Series(np.ones(len(uniq)), index=uniq)
        s_non_uniq = pd.Series(np.ones(len(non_uniq)), index=non_uniq)

        try:
            res = func(s_uniq, s_non_uniq)
        except Exception as e:
            rec["error_type"] = str(type(e))
            rec["error_message"] = str(e)
        records.append(rec)
    return records

funcs = [list_of_series_constructor, crosstab, concat_series, concat_dfs]

index_makers = [
    tm.makeStringIndex,
    tm.makeIntIndex,
    tm.makeUIntIndex,
    tm.makeFloatIndex,
    tm.makeDateIndex,
    tm.makeTimedeltaIndex,
    tm.makePeriodIndex,
    tm.makeMultiIndex,
    tm.makeBoolIndex,
]

pd.concat(
    {
        f.__name__: pd.DataFrame(duplicated_indices_errors(f, index_makers)) 
        for f in funcs
    }
).to_markdown()
```

|                                   | index_maker        | error_type                                | error_message                                            |
|:----------------------------------|:-------------------|:------------------------------------------|:---------------------------------------------------------|
| ('list_of_series_constructor', 0) | makeStringIndex    | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 1) | makeIntIndex       | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 2) | makeUIntIndex      | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 3) | makeFloatIndex     | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 4) | makeDateIndex      | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 5) | makeTimedeltaIndex | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 6) | makePeriodIndex    | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 7) | makeMultiIndex     | <class 'ValueError'>                      | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 8) | makeBoolIndex      | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('crosstab', 0)                   | makeStringIndex    | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 1)                   | makeIntIndex       | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 2)                   | makeUIntIndex      | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 3)                   | makeFloatIndex     | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 4)                   | makeDateIndex      | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 5)                   | makeTimedeltaIndex | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 6)                   | makePeriodIndex    | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 7)                   | makeMultiIndex     | <class 'ValueError'>                      | cannot handle a non-unique multi-index!                  |
| ('crosstab', 8)                   | makeBoolIndex      | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 0)              | makeStringIndex    | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 1)              | makeIntIndex       | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 2)              | makeUIntIndex      | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 3)              | makeFloatIndex     | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 4)              | makeDateIndex      | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 5)              | makeTimedeltaIndex | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 6)              | makePeriodIndex    | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 7)              | makeMultiIndex     | <class 'ValueError'>                      | cannot handle a non-unique multi-index!                  |
| ('concat_series', 8)              | makeBoolIndex      | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_dfs', 0)                 | makeStringIndex    | <class 'ValueError'>                      | Shape of passed values is (5, 2), indices imply (4, 2)   |
| ('concat_dfs', 1)                 | makeIntIndex       | <class 'ValueError'>                      | Shape of passed values is (7, 2), indices imply (5, 2)   |
| ('concat_dfs', 2)                 | makeUIntIndex      | <class 'ValueError'>                      | Shape of passed values is (7, 2), indices imply (5, 2)   |
| ('concat_dfs', 3)                 | makeFloatIndex     | <class 'ValueError'>                      | Shape of passed values is (7, 2), indices imply (5, 2)   |
| ('concat_dfs', 4)                 | makeDateIndex      | <class 'ValueError'>                      | Shape of passed values is (7, 2), indices imply (5, 2)   |
| ('concat_dfs', 5)                 | makeTimedeltaIndex | <class 'ValueError'>                      | Shape of passed values is (7, 2), indices imply (5, 2)   |
| ('concat_dfs', 6)                 | makePeriodIndex    | <class 'ValueError'>                      | Shape of passed values is (7, 2), indices imply (5, 2)   |
| ('concat_dfs', 7)                 | makeMultiIndex     | <class 'ValueError'>                      | cannot handle a non-unique multi-index!                  |
| ('concat_dfs', 8)                 | makeBoolIndex      | <class 'ValueError'>                      | Shape of passed values is (4, 2), indices imply (2, 2)   |

</details>


Instead of just giving concat a better error message, all these methods could get better (and more consistent) error messages. I think it's important for the rules about unions and intersections of indices to be defined in these more internal methods so behaviour isn't defined (and reimplemented) per top level function.

Additionally, this function essentially returns nonsense for unions when there are duplicated indices since different `Index` types have different definitions for unions with duplicates, so I think it's appropriate to throw errors here instead of passing on those nonsense results.

This has also made me realize there is one case where it's okay for there to be duplicates, which is when indices contain duplicates, but the intersection would not.

Reply:
> This has also made me realize there is one case where it's okay for there to be duplicates, which is when indices contain duplicates, but the intersection would not.

This got a bit more complicated. `idx.get_indexer(keys)` throws an error if `idx` is non-unique, regardless of whether `keys` have unique indices in `idx`.

Interestingly enough, it does not throw an error if the `idx` is non-unique, but `keys` is empty (which is the case I had initially seen).

Examples using pandas 1.1.5:

```python
pd.crosstab(
    pd.Series(np.arange(4), index=[0, 1, 1, 2]),
    pd.Series(np.arange(2), index=[0, 2]),
)
# ValueError: cannot reindex from a duplicate axis

pd.crosstab(
    pd.Series(np.arange(4), index=[0, 1, 1, 2]),
    pd.Series(np.arange(2), index=[3, 4]),
)
# Empty DataFrame
# Columns: []
# Index: []
```

<details>
<summary> Examples with pd.concat </summary>

This had the same behaviour despite not using `.get_indexer` directly in 1.1.5.

```python
pd.concat(
    [
        pd.Series(np.arange(4), index=[0, 1, 1, 2]),
        pd.Series(np.arange(2), index=[0, 2]),
    ],
    axis=1,
    join="inner"
)
# ValueError: cannot reindex from a duplicate axis

pd.concat(
    [
        pd.Series(np.arange(4), index=[0, 1, 1, 2]),
        pd.Series(np.arange(2), index=[3, 4]),
    ],
    axis=1,
    join="inner"
)
# Empty DataFrame
# Columns: [0, 1]
# Index: []
```

</details>

Reply:
Opened a more general issue for this: #38797